File: src/lib/scanner.ts
import fs from "fs";
import path from "path";

const VIDEO_EXTENSIONS = new Set([
  ".mkv", ".mp4", ".avi", ".mov", ".wmv", ".flv", ".m4v",
  ".mpg", ".mpeg", ".ts", ".m2ts", ".vob", ".iso", ".webm",
]);

const DEFAULT_SUBTITLE_EXTENSIONS = new Set([".srt"]);

const SUBTITLE_FOLDER_NAMES = new Set([
  "subs", "sub", "subtitles", "subtitle",
]);

/** Quick check for season/episode indicators in a filename */
const HAS_EPISODE_INFO = /[Ss]\d{1,2}[Ee]\d{1,3}|\d{1,2}[xX]\d{2,3}|[Ss]eason\s*\d{1,2}\s*[Ee]pisode/i;

const SEASON_FOLDER_PATTERN = /^(?:Season\s*|S)(\d+)$/i;
const SPECIALS_FOLDER_NAMES = new Set(["specials", "season 0", "season 00", "season0", "season00"]);

const EXTRA_FOLDER_MAP: Record<string, string> = {
  "extras": "other",
  "extra": "other",
  "behind the scenes": "behind_the_scenes",
  "behindthescenes": "behind_the_scenes",
  "deleted scenes": "deleted_scenes",
  "deletedscenes": "deleted_scenes",
  "featurettes": "featurettes",
  "featurette": "featurettes",
  "interviews": "interviews",
  "interview": "interviews",
  "scenes": "scenes",
  "scene": "scenes",
  "shorts": "shorts",
  "short": "shorts",
  "trailers": "trailers",
  "trailer": "trailers",
  "other": "other",
};

export interface ScannedFile {
  sourcePath: string;
  fileName: string;
  fileSize: number;
  fileExtension: string;
}

export type FileCategory = "episode" | "movie" | "special" | "extra";

export interface ScannedGroupFile {
  sourcePath: string;
  fileName: string;
  fileSize: number;
  fileExtension: string;
  detectedSeason: number | null;
  fileCategory: FileCategory;
  extraType: string | null;
}

export interface ScannedSubtitleFile {
  sourcePath: string;
  fileName: string;
  fileExtension: string;
  languageCode: string | null;
  /** The base name used for matching to a video file (without language code and subtitle extension) */
  matchBase: string;
}

export interface ScannedGroup {
  folderPath: string;
  folderName: string;
  files: ScannedGroupFile[];
  subtitles: ScannedSubtitleFile[];
}

/**
 * Parse a subtitle filename to extract the language code and match base.
 * E.g. "Show.S01E01.en.srt" → { languageCode: "en", matchBase: "show.s01e01" }
 *      "Show.S01E01.srt"    → { languageCode: null, matchBase: "show.s01e01" }
 */
function parseSubtitleName(
  fileName: string,
  subtitleExts: Set<string>
): { languageCode: string | null; matchBase: string } | null {
  const ext = path.extname(fileName).toLowerCase();
  if (!subtitleExts.has(ext)) return null;

  const withoutExt = fileName.slice(0, -ext.length);
  // Check if the part before .srt is a 2-3 char language code
  const lastDot = withoutExt.lastIndexOf(".");
  if (lastDot !== -1) {
    const possibleLang = withoutExt.slice(lastDot + 1);
    if (/^[a-z]{2,3}$/i.test(possibleLang)) {
      return {
        languageCode: possibleLang.toLowerCase(),
        matchBase: withoutExt.slice(0, lastDot).toLowerCase(),
      };
    }
  }
  return { languageCode: null, matchBase: withoutExt.toLowerCase() };
}

function collectSubtitleFiles(
  dir: string,
  subtitleExts: Set<string>,
  languageFilter: Set<string> | null
): ScannedSubtitleFile[] {
  const results: ScannedSubtitleFile[] = [];
  let entries: fs.Dirent[];
  try {
    entries = fs.readdirSync(dir, { withFileTypes: true });
  } catch {
    return results;
  }
  for (const entry of entries) {
    const fullPath = path.join(dir, entry.name);
    if (entry.isDirectory()) {
      results.push(...collectSubtitleFiles(fullPath, subtitleExts, languageFilter));
    } else if (entry.isFile()) {
      const ext = path.extname(entry.name).toLowerCase();
      if (subtitleExts.has(ext)) {
        const parsed = parseSubtitleName(entry.name, subtitleExts);
        if (parsed) {
          // Language filter: if filter is set and subtitle has a language code, it must match
          if (languageFilter && parsed.languageCode && !languageFilter.has(parsed.languageCode)) {
            continue;
          }
          results.push({
            sourcePath: fullPath,
            fileName: entry.name,
            fileExtension: ext,
            languageCode: parsed.languageCode,
            matchBase: parsed.matchBase,
          });
        }
      }
    }
  }
  return results;
}

function collectVideoFiles(dir: string): ScannedFile[] {
  const results: ScannedFile[] = [];
  const entries = fs.readdirSync(dir, { withFileTypes: true });
  for (const entry of entries) {
    const fullPath = path.join(dir, entry.name);
    if (entry.isDirectory()) {
      results.push(...collectVideoFiles(fullPath));
    } else if (entry.isFile()) {
      const ext = path.extname(entry.name).toLowerCase();
      if (VIDEO_EXTENSIONS.has(ext)) {
        const stat = fs.statSync(fullPath);
        results.push({
          sourcePath: fullPath,
          fileName: entry.name,
          fileSize: stat.size,
          fileExtension: ext,
        });
      }
    }
  }
  return results;
}

function classifySubfolder(
  folderName: string
): {
  detectedSeason: number | null;
  fileCategory: FileCategory;
  extraType: string | null;
} {
  const lower = folderName.toLowerCase().trim();

  // Check specials
  if (SPECIALS_FOLDER_NAMES.has(lower)) {
    return { detectedSeason: 0, fileCategory: "special", extraType: null };
  }

  // Check season pattern
  const seasonMatch = folderName.match(SEASON_FOLDER_PATTERN);
  if (seasonMatch) {
    const seasonNum = parseInt(seasonMatch[1], 10);
    if (seasonNum === 0) {
      return { detectedSeason: 0, fileCategory: "special", extraType: null };
    }
    return { detectedSeason: seasonNum, fileCategory: "episode", extraType: null };
  }

  // Check extras
  const extraType = EXTRA_FOLDER_MAP[lower];
  if (extraType) {
    return { detectedSeason: null, fileCategory: "extra", extraType };
  }

  return { detectedSeason: null, fileCategory: "episode", extraType: null };
}

export function scanDirectoryGrouped(
  dirPath: string,
  subtitleExtensions?: Set<string>,
  subtitleLanguages?: Set<string> | null
): ScannedGroup[] {
  const subExts = subtitleExtensions ?? DEFAULT_SUBTITLE_EXTENSIONS;
  const langFilter = subtitleLanguages && subtitleLanguages.size > 0 ? subtitleLanguages : null;
  const groups: ScannedGroup[] = [];
  const entries = fs.readdirSync(dirPath, { withFileTypes: true });

  for (const entry of entries) {
    const fullPath = path.join(dirPath, entry.name);

    if (entry.isDirectory()) {
      const group: ScannedGroup = {
        folderPath: fullPath,
        folderName: entry.name,
        files: [],
        subtitles: [],
      };

      // Walk the group folder
      const subEntries = fs.readdirSync(fullPath, { withFileTypes: true });
      let hasSeasonFolders = false;

      for (const sub of subEntries) {
        const subPath = path.join(fullPath, sub.name);

        if (sub.isDirectory()) {
          // Check if this is a subtitle folder
          if (SUBTITLE_FOLDER_NAMES.has(sub.name.toLowerCase().trim())) {
            group.subtitles.push(...collectSubtitleFiles(subPath, subExts, langFilter));
            continue;
          }

          const classification = classifySubfolder(sub.name);
          if (classification.fileCategory === "episode" && classification.detectedSeason !== null) {
            hasSeasonFolders = true;
          }

          // Collect all video files in this subfolder
          const files = collectVideoFiles(subPath);
          for (const file of files) {
            group.files.push({
              ...file,
              detectedSeason: classification.detectedSeason,
              fileCategory: classification.fileCategory,
              extraType: classification.extraType,
            });
          }

          // Also collect subtitles from season/extra subfolders and their sub dirs
          group.subtitles.push(...collectSubtitleFiles(subPath, subExts, langFilter));
        } else if (sub.isFile()) {
          const ext = path.extname(sub.name).toLowerCase();
          if (VIDEO_EXTENSIONS.has(ext)) {
            const stat = fs.statSync(subPath);
            group.files.push({
              sourcePath: subPath,
              fileName: sub.name,
              fileSize: stat.size,
              fileExtension: ext,
              detectedSeason: null, // will be inferred from filename later
              fileCategory: "episode", // default, may be reclassified
              extraType: null,
            });
          } else if (subExts.has(ext)) {
            // Subtitle file at group root level
            const parsed = parseSubtitleName(sub.name, subExts);
            if (parsed) {
              // Language filter: if filter is set and subtitle has a language code, it must match
              if (!langFilter || !parsed.languageCode || langFilter.has(parsed.languageCode)) {
                group.subtitles.push({
                  sourcePath: subPath,
                  fileName: sub.name,
                  fileExtension: ext,
                  languageCode: parsed.languageCode,
                  matchBase: parsed.matchBase,
                });
              }
            }
          }
        }
      }

      if (group.files.length > 0) {
        // Media type heuristic: if season folders exist or multiple files → TV
        // Single file with no season folders and no episode info in filename → movie
        if (!hasSeasonFolders && group.files.length === 1 &&
            group.files.every((f) => f.fileCategory === "episode") &&
            !HAS_EPISODE_INFO.test(group.files[0].fileName)) {
          group.files[0].fileCategory = "movie";
        }

        groups.push(group);
      }
    } else if (entry.isFile()) {
      // Loose file in scan root → single-file group
      const ext = path.extname(entry.name).toLowerCase();
      if (VIDEO_EXTENSIONS.has(ext)) {
        const stat = fs.statSync(fullPath);
        // Detect if the filename contains season/episode info → TV, otherwise movie
        const hasEpisodeInfo = HAS_EPISODE_INFO.test(entry.name);
        groups.push({
          folderPath: dirPath,
          folderName: entry.name.replace(/\.[^.]+$/, ""),
          files: [
            {
              sourcePath: fullPath,
              fileName: entry.name,
              fileSize: stat.size,
              fileExtension: ext,
              detectedSeason: null,
              fileCategory: hasEpisodeInfo ? "episode" : "movie",
              extraType: null,
            },
          ],
          subtitles: [],
        });
      }
    }
  }

  return groups;
}

// Keep legacy flat scanner for backward compatibility
export function scanDirectory(dirPath: string): ScannedFile[] {
  return collectVideoFiles(dirPath);
}

File: src-tauri/src/scanner.rs
use once_cell::sync::Lazy;
use std::collections::{HashMap, HashSet};
use std::fs;
use std::path::Path;
use tracing::warn;

static VIDEO_EXTENSIONS: Lazy<HashSet<&'static str>> = Lazy::new(|| {
    [
        ".mkv", ".mp4", ".avi", ".mov", ".wmv", ".flv", ".m4v", ".mpg", ".mpeg", ".ts", ".m2ts",
        ".vob", ".iso", ".webm",
    ]
    .into_iter()
    .collect()
});

static SPECIALS_FOLDER_NAMES: Lazy<HashSet<&'static str>> = Lazy::new(|| {
    ["specials", "season 0", "season 00", "season0", "season00"]
        .into_iter()
        .collect()
});

static EXTRA_FOLDER_MAP: Lazy<HashMap<&'static str, &'static str>> = Lazy::new(|| {
    [
        ("extras", "other"),
        ("extra", "other"),
        ("behind the scenes", "behind_the_scenes"),
        ("behindthescenes", "behind_the_scenes"),
        ("deleted scenes", "deleted_scenes"),
        ("deletedscenes", "deleted_scenes"),
        ("featurettes", "featurettes"),
        ("featurette", "featurettes"),
        ("interviews", "interviews"),
        ("interview", "interviews"),
        ("scenes", "scenes"),
        ("scene", "scenes"),
        ("shorts", "shorts"),
        ("short", "shorts"),
        ("trailers", "trailers"),
        ("trailer", "trailers"),
        ("other", "other"),
    ]
    .into_iter()
    .collect()
});

static SEASON_FOLDER_RE: Lazy<regex::Regex> =
    Lazy::new(|| regex::Regex::new(r"(?i)^(?:Season\s*|S)(\d+)$").unwrap());

#[derive(Debug, Clone)]
pub struct ScannedFile {
    pub source_path: String,
    pub file_name: String,
    pub file_size: i64,
    pub file_extension: String,
}

#[derive(Debug, Clone)]
pub struct ScannedGroupFile {
    pub source_path: String,
    pub file_name: String,
    pub file_size: i64,
    pub file_extension: String,
    pub detected_season: Option<i64>,
    pub file_category: String,
    pub extra_type: Option<String>,
}

#[derive(Debug, Clone)]
pub struct ScannedGroup {
    pub folder_path: String,
    pub folder_name: String,
    pub files: Vec<ScannedGroupFile>,
}

struct SubfolderClassification {
    detected_season: Option<i64>,
    file_category: String,
    extra_type: Option<String>,
}

fn classify_subfolder(folder_name: &str) -> SubfolderClassification {
    let lower = folder_name.to_lowercase().trim().to_string();

    // Check specials
    if SPECIALS_FOLDER_NAMES.contains(lower.as_str()) {
        return SubfolderClassification {
            detected_season: Some(0),
            file_category: "special".into(),
            extra_type: None,
        };
    }

    // Check season pattern
    if let Some(caps) = SEASON_FOLDER_RE.captures(folder_name) {
        if let Ok(num) = caps[1].parse::<i64>() {
            if num == 0 {
                return SubfolderClassification {
                    detected_season: Some(0),
                    file_category: "special".into(),
                    extra_type: None,
                };
            }
            return SubfolderClassification {
                detected_season: Some(num),
                file_category: "episode".into(),
                extra_type: None,
            };
        }
    }

    // Check extras
    if let Some(&extra_type) = EXTRA_FOLDER_MAP.get(lower.as_str()) {
        return SubfolderClassification {
            detected_season: None,
            file_category: "extra".into(),
            extra_type: Some(extra_type.into()),
        };
    }

    SubfolderClassification {
        detected_season: None,
        file_category: "episode".into(),
        extra_type: None,
    }
}

fn collect_video_files(dir: &Path) -> Vec<ScannedFile> {
    let mut results = Vec::new();
    let entries = match fs::read_dir(dir) {
        Ok(e) => e,
        Err(e) => {
            warn!("Cannot read directory {}: {}", dir.display(), e);
            return results;
        }
    };

    for entry in entries.flatten() {
        let path = entry.path();
        if path.is_dir() {
            results.extend(collect_video_files(&path));
        } else if path.is_file() {
            let ext = path
                .extension()
                .and_then(|e| e.to_str())
                .map(|e| format!(".{}", e.to_lowercase()))
                .unwrap_or_default();
            if VIDEO_EXTENSIONS.contains(ext.as_str()) {
                let size = fs::metadata(&path).map(|m| m.len() as i64).unwrap_or(0);
                results.push(ScannedFile {
                    source_path: path.to_string_lossy().to_string(),
                    file_name: entry.file_name().to_string_lossy().to_string(),
                    file_size: size,
                    file_extension: ext,
                });
            }
        }
    }
    results
}

pub fn scan_directory_grouped(dir_path: &str) -> Vec<ScannedGroup> {
    let dir = Path::new(dir_path);
    let mut groups = Vec::new();

    let entries = match fs::read_dir(dir) {
        Ok(e) => e,
        Err(e) => {
            warn!("Cannot read scan directory {}: {}", dir_path, e);
            return groups;
        }
    };

    for entry in entries.flatten() {
        let path = entry.path();
        let name = entry.file_name().to_string_lossy().to_string();

        if path.is_dir() {
            let mut group = ScannedGroup {
                folder_path: path.to_string_lossy().to_string(),
                folder_name: name,
                files: Vec::new(),
            };

            let sub_entries = match fs::read_dir(&path) {
                Ok(e) => e,
                Err(_) => continue,
            };

            let mut has_season_folders = false;

            for sub in sub_entries.flatten() {
                let sub_path = sub.path();
                let sub_name = sub.file_name().to_string_lossy().to_string();

                if sub_path.is_dir() {
                    let classification = classify_subfolder(&sub_name);
                    if classification.file_category == "episode"
                        && classification.detected_season.is_some()
                    {
                        has_season_folders = true;
                    }

                    let files = collect_video_files(&sub_path);
                    for file in files {
                        group.files.push(ScannedGroupFile {
                            source_path: file.source_path,
                            file_name: file.file_name,
                            file_size: file.file_size,
                            file_extension: file.file_extension,
                            detected_season: classification.detected_season,
                            file_category: classification.file_category.clone(),
                            extra_type: classification.extra_type.clone(),
                        });
                    }
                } else if sub_path.is_file() {
                    let ext = sub_path
                        .extension()
                        .and_then(|e| e.to_str())
                        .map(|e| format!(".{}", e.to_lowercase()))
                        .unwrap_or_default();
                    if VIDEO_EXTENSIONS.contains(ext.as_str()) {
                        let size = fs::metadata(&sub_path)
                            .map(|m| m.len() as i64)
                            .unwrap_or(0);
                        group.files.push(ScannedGroupFile {
                            source_path: sub_path.to_string_lossy().to_string(),
                            file_name: sub_name,
                            file_size: size,
                            file_extension: ext,
                            detected_season: None,
                            file_category: "episode".into(),
                            extra_type: None,
                        });
                    }
                }
            }

            if !group.files.is_empty() {
                // Media type heuristic: single file, no season folders → movie
                if !has_season_folders
                    && group.files.len() == 1
                    && group.files.iter().all(|f| f.file_category == "episode")
                {
                    group.files[0].file_category = "movie".into();
                }
                groups.push(group);
            }
        } else if path.is_file() {
            // Loose file in scan root → single-file group (movie)
            let ext = path
                .extension()
                .and_then(|e| e.to_str())
                .map(|e| format!(".{}", e.to_lowercase()))
                .unwrap_or_default();
            if VIDEO_EXTENSIONS.contains(ext.as_str()) {
                let size = fs::metadata(&path).map(|m| m.len() as i64).unwrap_or(0);
                let stem = path
                    .file_stem()
                    .and_then(|s| s.to_str())
                    .unwrap_or("")
                    .to_string();
                groups.push(ScannedGroup {
                    folder_path: dir_path.to_string(),
                    folder_name: stem,
                    files: vec![ScannedGroupFile {
                        source_path: path.to_string_lossy().to_string(),
                        file_name: name,
                        file_size: size,
                        file_extension: ext,
                        detected_season: None,
                        file_category: "movie".into(),
                        extra_type: None,
                    }],
                });
            }
        }
    }

    groups
}
